### ISSUE TITLE
Improve memory consumption in MetaModelica by allocating a new component just once

### ISSUE THREAD

**Issue by adrpo** [OpenModelica/OpenModelica#11909]
Currently we have a lot of patterns like:
```mo
match c 
  case C
     c.c1 = someValue;
     c.c2 = someValue;
```

If  `c ` component is very large we allocate that size at each statement that sets a subcomponent of it.
Do some analysis and change the code to:
```mo
c = RECORD(c1, c2)
```

**Comment by perost** [comment]
Duplicate of [#3814](https://trac.openmodelica.org/OpenModelica/ticket/3814).

**Comment by adrpo** [comment]
Not really a duplicate of [#3814](https://trac.openmodelica.org/OpenModelica/ticket/3814), that one proposes alternative syntax, I just want to gather all assignments to the component subcomponents, allocate once and then set them once.

**Comment by JKRT** [comment]
I notified Adrian when I profiled the Julia variant at some places we did much more allocations than we should have for some branches. The reason is that Julia currently works as MetaModelica, making one allocation for each write to a subcomponent of an immutable datatype; hence, sometimes twice the amount of allocations are needed. 

I managed to get memory down by 10-15% by changing this at a few places. 

I think one could expect similar results for MetaModelica, or even more if one implements this. I guess one could look at the basic blocks (Statement Lists) and see if more than one assignment is done to an immutable within that block. If that is the case we rewrite it to call the object constructor directly with the fields updated.

**Comment by hkiel** [comment]
@adrpo I opened a PR addressing this issue. Is it worth changing old and Backend old Frontend code?

**Comment by phannebohm** [comment]
Sorry is the idea to do that manually? It would be nice if the compiler did it for us. Especially when we have things like below so we don't have to write all that stuff, also for readability and maintainability reasons.
```modelica
match c 
  case RECORD(a, b, c, ..., x, y, z) then RECORD(new_a, b, c, ..., x, new_y, z);
```

**Comment by adrpo** [comment]
Yes, we should fix the compiler to do that. We would need to collect all assignments done to the same component parent in the function and group them into one. I will check if we can do this somewhere in the backend or the simcodegen.

**Comment by casella** [comment]
Good if we manage to bring down the memory consumption of MetaModelica 🥳

**Comment by JKRT** [comment]
Attempting to solve this in https://github.com/OpenModelica/OpenModelica/pull/15396 looks promising in terms of improved performance and reduced memory consumption

**Comment by casella** [comment]
@JKRT do you expect major memory savings somewhere?

**Comment by JKRT** [comment]
@casella Sorry I missed this comment. In many instances in both the NF and the NB people who know about this issue have been good at optimizing object-creation by hand to work around this issue; this fix should improve memory use for many paths in the old backend and if nothing else do so for new code:)
